src/openstack_cloud/openstack_manager.py

```diff
@@ -777,28 +783,45 @@ def get_github_runner_info(self) -> tuple[RunnerGithubInfo]:
             if runner.name.startswith(f"{self.instance_name}-")
         )
 
-    def _get_ssh_connections(self, instance: Server) -> Generator[SshConnection, None, None]:
+    def _get_ssh_connections(self, server: Server) -> Generator[SshConnection, None, None]:
         """Get ssh connections within a network for a given openstack instance.
 
         Args:
-            instance: The Openstack server instance.
+            server: The Openstack server instance.
 
         Yields:
-            Openstack SSH connections.
+            SSH connections to OpenStack server instance.
         """
-        if not cast(dict, instance.addresses).get(self._config.network, None):
-            logger.error(
+        if not server.key_name:
+            logger.warning(
+                "Unable to create SSH connection as no valid keypair found for %s", server.name
+            )
+            return
+
+        if not cast(dict, server.addresses).get(self._config.network, None):
+            logger.warning(
                 "Instance %s created under invalid or no network %s",
-                instance.instance_name,
-                instance.addresses,
+                server.name,
+                server.addresses,
             )
             return
-        for address in instance.addresses[self._config.network]:
+
+        for address in server.addresses[self._config.network]:
             ip = address["addr"]
+
+            key_path = self._get_key_path(server.name)
+            if not key_path.exists():
+                logger.warning(
```

Comment:
I'm wondering if this log could be promoted to an error log since human intervention might be needed (unrecoverable)